sql/wsrep_sst.cc

```diff
@@ -1465,121 +1438,313 @@ std::string wsrep_sst_prepare()
+      ret= err;
     }
+    thd->m_statement_psi= parent_locker;
+    thd->end_statement();
+    close_thread_tables(thd);
+    delete_explain_query(thd->lex);
   }
 
-  return -ret;
+  thd->reset_query();
+  thd->cleanup_after_query();
+  return ret;
 }
 
-static void sst_reject_queries(my_bool close_conn)
+
+#define SST_USER_PREFIX "wsrep.sst."
+
+static std::string
+generate_user_name(const char* const node_name)
 {
-  WSREP_INFO("Rejecting client queries for the duration of SST.");
-  if (TRUE == close_conn) wsrep_close_client_connections(FALSE);
+  static size_t const max_user_name_length(32); // MySQL 8.0
+
+  struct timeval tv;
+  gettimeofday(&tv, NULL);
+  time_t time_now;
+  time_now = tv.tv_sec;
+  struct tm tm_now;
+  (void)gmtime_r(&time_now, &tm_now);
+  char timestamp[32];
+  strftime(timestamp, sizeof(timestamp), "%y%m%d_%H%M%S", &tm_now);
+
+  return (std::string(SST_USER_PREFIX) + timestamp + "_" + node_name).substr
+          (0, max_user_name_length);
 }
 
-static int sst_donate_mysqldump (const char*         addr,
-                                 const wsrep::gtid&  gtid,
-                                 bool                bypass,
-                                 char**              env) // carries auth info
+void
+wsrep_sst_cleanup_user(THD* const thd)
 {
-  char host[256];
-  wsp::Address address(addr);
-  if (!address.is_valid())
-  {
-    WSREP_ERROR("Could not parse SST address : %s", addr);
-    return 0;
+  int err(run_sql_command(
+          thd, "DELETE FROM mysql.user WHERE user LIKE '" SST_USER_PREFIX "%';"));
+  if (err && err != ER_NO_SUCH_TABLE) {
+    WSREP_WARN("Failed to clean up SST user(s): %d", err);
   }
-  memcpy(host, address.get_address(), address.get_address_len());
-  int port= address.get_port();
-  bool extra_args;
-  size_t const cmd_len= estimate_cmd_len(&extra_args);
-  wsp::string cmd_str(cmd_len);
+}
 
-  if (!cmd_str())
+static std::string
+generate_password(int size) {
+  // Password characters are limited, because we are using these passwords
+  // within a bash script.
+  static const std::string g_allowed_pwd_chars(
+    "qwertyuiopasdfghjklzxcvbnm1234567890"
+    "QWERTYUIOPASDFGHJKLZXCVBNM");
+  // To guarantee MySQL password requirements: upper/lowercase, numbers,
+  // special characters
+  const char* const pwd_prefix("yx9!A-");
+
+  std::stringstream ss;
+  ss << pwd_prefix;
+  size -= strlen(pwd_prefix);
+
+  ulong const seed1= (ulong) (my_rnd(&sql_rand) * 0xffffffff);
+  ulong const seed2= (ulong) time((time_t*) 0);
+  struct my_rnd_struct rand;
+  my_rnd_init(&rand, seed1 + (ulong)((size_t) &rand), seed2);
+
+  while (size > 0) {
+    int ch = ((int)(my_rnd(&rand) * 0xffff)) % g_allowed_pwd_chars.size();
+    ss << g_allowed_pwd_chars[ch];
+    --size;
+  }
+
+  return ss.str();
+}
+
+static const char**
+sst_user_grants(const std::string& method)
+{
+  static const char* mariabackup[] = {
+    "GRANT RELOAD, PROCESS, LOCK TABLES, BINLOG MONITOR ON *.* TO '%s'@'localhost';",
```

Comment:
As this is a serious fault I've added https://github.com/MariaDB/mariadb-docker/commit/3fb9801bbfa5768439dbc4e001af10fd06d72ac2 as a test.

Reply:
This is a localhost connection. The script is being launched locally on donor. What is exactly a fault here? And why would we use a donor IP instead of a socket?